[RA-1 Ch02]  Requirmeents and Recommendations (#1223)
cntt-n/CNTT: doc/ref_arch/openstack/chapters/chapter02.md

  * [2.3.6 LCM Requirements](#2.3.6)
  * [2.3.7 Assurance Requirements](#2.3.7)
  * [2.3.8 Security Requirements](#2.3.8)

> * [2.4 Architecture and OpenStack Recommendations.](#2.4)

Reviewer: No period at the end.

Author: @deebhatia Done

  * [2.4.1 General Requirements](#2.4.1)
  * [2.4.2 Infrastructure Requirements](#2.4.2)
  * [2.4.3 VIM Requirements](#2.4.3)
  * [2.4.4 Interfaces & APIs Requirements](#2.4.4)
  * [2.4.5 Tenant Requirements](#2.4.5)
  * [2.4.6 LCM Requirements](#2.4.6)
  * [2.4.7 Assurance Requirements](#2.4.7)

>   * [2.4.8 Security Requirements](#2.4.8)

Reviewer: Requirements -> Recommendations

Author: @deebhatia Thanks -- Done

|----|---------------|---------------------|---------------|
| `req.gen.ost.01` | Open source | The Architecture **must** use OpenStack APIs.| [RA-1 5.3](./chapter05.md#53-consolidated-set-of-apis) |

> | `req.gen.ost.02` | Open source | The Architecture **must** support dynamic request and configuration of virtual resources (compute, network, storage) through OpenStack APIs. | [RA-1 5.3](./chapter05.md#53-consolidated-set-of-apis) |

Reviewer 2: Sequence numbers needs fixings

Author: @rabi-abdel the numbering reflects the sub-category; this practice is inherited for RM

> | `req.inf.stg.02` | Storage | The Architecture **should** provide Object storage for VM Instances. | OpenStack Swift Service ([RA-1 4.3.1.4 "Swift"](./chapter04.md#4314-swift)) |

Reviewer 3: From my understanding, Swift is mandatory in RA1 Chapter 5 and then it may be fixed here and switch to **must**.
Setting as optional must be discussed all together.

Reviewer 4: A RM change #524 just went through to mark Object Storage services as optional.

| `req.sec.ntw.04` | Networking | The Architecture **must** configure all of the underlay network components to ensure the complete separation from the overlay customer deployments. | |
| `req.sec.ntw.05` | Networking | The Architecture **must** have the underlay network include strong access controls that adhere to the V1.1 NIST Cybersecurity Framework. | |
<!-- | `req.sec.ntw.01` | Networking | The Architecture **must** have the underlay network include strong access controls that comply with the applicable security standard (national, regional), for example ISO27001. | |
| `req.sec.ntw.02` | Networking | The Architecture **must** have all security logs stored in accordance with applicable security standard (national, regional), for example ISO27001. | | -->

<p align="center"><b>Table 2-8:</b> OpenStack Security Requirements.</p>

<a name="2.4"></a>
### 2.4 Architecture and OpenStack Recommendations

The requirements listed in this section are optional, viz. do not need to be implemented in conformant implementations.

2.4.1 General Recommendations

| Ref # | sub-category | Description |  Traceability |
|----|-------|---------------------|-------------|

> | `req.gen.cnt.01` | Cloud nativeness | The Architecture **should** consist of stateless service components. However, where state is required it must be kept external to the component. | OpenStack consists of both stateless and stateful services where the stateful services utilize a database. For latter see "[Configuring the stateful services](https://docs.openstack.org/ha-guide/control-plane-stateful.html)"|

Reviewer: __Query__ - @pgoyal01 Should we use `rec.gen.cnt.01` instead of `req.gen.cnt.01` ? A person might end up incrementing the code in requirement section which is already being utilised here or do you want to control at process level?

Author: @deebhatia Thanks -- Done. Didn't re-sequence the numbering.

| `req.sec.ntw.03` | Networking | The Architecture **must** have the underlay network incorporate encrypted and/or private communications channels to ensure its security. | |
| `req.sec.ntw.04` | Networking | The Architecture **must** configure all of the underlay network components to ensure the complete separation from the overlay customer deployments. | |
| `req.sec.ntw.05` | Networking | The Architecture **must** have the underlay network include strong access controls that adhere to the V1.1 NIST Cybersecurity Framework. | |
<!-- | `req.sec.ntw.01` | Networking | The Architecture **must** have the underlay network include strong access controls that comply with the applicable security standard (national, regional), for example ISO27001. | |
| `req.sec.ntw.02` | Networking | The Architecture **must** have all security logs stored in accordance with applicable security standard (national, regional), for example ISO27001. | | -->

<p align="center"><b>Table 2-8:</b> OpenStack Security Requirements.</p>

<a name="2.4"></a>
### 2.4 Architecture and OpenStack Recommendations

> The requirements listed in this section are optional, viz. do not need to be implemented in conformant implementations.

Reviewer 4: ```suggestion
The requirements listed in this section are optional, and are not required in order to be deemed a conformant implementation.
```
slight edit; not sure about the use of 'viz.' in this sentence.

Author: @kedmison Changed - Thanks

<a name="2.4.1"></a>
2.4.1 General Recommendations

| Ref # | sub-category | Description |  Traceability |
|----|-------|---------------------|-------------|

> | `rec.gen.cnt.01` | Cloud nativeness | The Architecture **should** consist of stateless service components. However, where state is required it must be kept external to the component. | OpenStack consists of both stateless and stateful services where the stateful services utilize a database. For latter see "[Configuring the stateful services](https://docs.openstack.org/ha-guide/control-plane-stateful.html)"|

Reviewer 4: I am a little concerned about how close 'rec.gen.*' is to 'req.gen.*' and the potential for inadvertent mistakes.  Can we use just one, req.gen?
* no inadvertent lexicographical mistakes
* no embedding of requirement strength ('rec'comendation vs 'req'uirement) in the naming, so if we change a requirement from a must to a should, or vice versa, we don't need to renumber?

Author: @karinesevilla had suggested using "rec". @kedmison suggested using "req" labeling even for Recommendations as was originally the case. 

In addition to advantages stated by @kedmison, using "req" has the advantage that none of the other RA-1 chapters have to edited. 

So naming has been reverted to use of "req"

| `rec.gen.cnt.02` | Cloud nativeness | The Architecture **should** consist of service components implemented as microservices that are individually dynamically scalable. | |

> | `rec.gen.scl.01` | Scalability | The Architecture **should** support policy driven auto-scaling. | This requirement is currently not addressed but will likley be supported through [Senlin](https://docs.openstack.org/senlin/pike/index.html), cluster management service. |

Reviewer 4: ```suggestion
| `rec.gen.scl.01` | Scalability | The Architecture **should** support policy driven auto-scaling. | This requirement is currently not addressed but will likely be supported through [Senlin](https://docs.openstack.org/senlin/pike/index.html), cluster management service. |
```

Author: @kedmison Corrected - Thanks

<p align="center"><b>Table 2-9:</b> General Recommendations</p>

<a name="2.4.2"></a>
2.4.2 Infrastructure Recommendations

| Ref # | sub-category | Description |  Traceability |
|----|-------|---------------------|-------------|
| `rec.inf.com.02` | Compute | The Architecture **should** include industry standard hardware management systems at both HW device level (embedded) and HW platform level (external to device). | |
| `rec.inf.com.03` | Compute | The Architecture **should** support symmetrical CPU multi-processing with shared memory access as well as multi-threading. | |
| `rec.inf.stg.05` | Storage | The Architecture **should** be able to accommodate VNFs that store back into its image through use of hypervisor attached volumes. | **Wouldn't that be a Security Violation** |
| `rec.inf.stg.08` | Storage | The Architecture **should** allow use of externally provided large archival storage for its Backup / Restore / Archival needs. | |
| `rec.inf.stg.09` | Storage | The Architecture **should** make available all non-host OS / Hypervisor / Host systems storage as network-based Block, File or Object Storage for tenant/management consumption. | |

> | `rec.inf.ntw.04` | Network | The Architecture **should** support service function chaining. | **Is this VNFM/NFVO requirement** |

Reviewer 4: Can we leave this out entirely until the networking discussions are concluded?

Author: @kedmison Since these are only recommendations, I am not sure if there is any harm in leaving them in place. Some of these may be lost as they may be not be the subject of the networking decisions.

| `rec.gen.cnt.02` | Cloud nativeness | The Architecture **should** consist of service components implemented as microservices that are individually dynamically scalable. | |
| `rec.gen.scl.01` | Scalability | The Architecture **should** support policy driven auto-scaling. | This requirement is currently not addressed but will likley be supported through [Senlin](https://docs.openstack.org/senlin/pike/index.html), cluster management service. |
| `rec.gen.rsl.02` | Resiliency | The Architecture **should** support resilient OpenStack service components that are not subject to `req.gen.rsl.01`. | |

<p align="center"><b>Table 2-9:</b> General Recommendations</p>

<a name="2.4.2"></a>
2.4.2 Infrastructure Recommendations

| Ref # | sub-category | Description |  Traceability |
|----|-------|---------------------|-------------|
| `rec.inf.com.02` | Compute | The Architecture **should** include industry standard hardware management systems at both HW device level (embedded) and HW platform level (external to device). | |
| `rec.inf.com.03` | Compute | The Architecture **should** support symmetrical CPU multi-processing with shared memory access as well as multi-threading. | |

> | `rec.inf.stg.05` | Storage | The Architecture **should** be able to accommodate VNFs that store back into its image through use of hypervisor attached volumes. | **Wouldn't that be a Security Violation** |

Reviewer 5: Can we remove this requirement (rec.inf.stg.05)? It is not only a security problem. It is against NFV architecture to allow a VNF write into the package repository and btw. the packages are encrypted. Also, how to deal with multiple instances of such VNF?
Why would anybody need that?

Author: @ulikleber Glad to do it. Thanks
